pybossa/view/account.py

```diff
@@ -1151,6 +1151,73 @@ def add_metadata(name):
     flash("Input saved successfully", "info")
     return redirect(url_for('account.profile', name=name))
 
+
+@blueprint.route('/<user_name>/taskbrowse_bookmarks/<short_name>', methods=['GET', 'POST'])
+@login_required
+def taskbrowse_bookmarks(user_name, short_name):
+    if current_user.name != user_name:
+        return abort(403)
+
+    # get all bookmarks for project
+    if request.method == 'GET':
+        taskbrowse_bookmarks = cached_users.get_taskbrowse_bookmarks(user_name)
+        proj_bookmarks = taskbrowse_bookmarks.get(short_name, {})
+        return jsonify(proj_bookmarks)
+    # add a bookmark
+    if request.method == 'POST':
+        user = user_repo.get_by_name(name=user_name)
+        taskbrowse_bookmarks = user.info.get('taskbrowse_bookmarks', {})
+        proj_bookmarks = taskbrowse_bookmarks.get(short_name, {})
+
+        MAX_BOOKMARK_NAME_LEN = 100
+        MAX_BOOKMARK_URL_LEN = 500
+
+        url = request.body.get('url', None)
+        name = request.body.get('name', None)
+
+        if name is None or len(name) > MAX_BOOKMARK_NAME_LEN:
+            error = f'Bookmark name must be between 1-{MAX_BOOKMARK_NAME_LEN} characters.'
+            current_app.logger.exception(f'Bad request: {error},  project: {short_name}, bookmark_name:{name}')
+            return Response(json.dumps({"description": error}), 400, mimetype='application/json')
+        if url is None or len(url) > MAX_BOOKMARK_URL_LEN:
+            error = 'Bookmark URL must be between 1-100 characters.'
+            current_app.logger.exception(f'Bad request: {error}, project: {short_name}, bookmark_url:{url}')
+            return Response(json.dumps({"description": error}), 400, mimetype='application/json')
+
+        proj_bookmarks[name] =  url
+        taskbrowse_bookmarks[short_name] = proj_bookmarks
+        user.info['taskbrowse_bookmarks'] = taskbrowse_bookmarks
+        user_repo.update(user)
+        cached_users.delete_taskbrowse_bookmarks(user)
+        return Response(json.dumps(proj_bookmarks), 200, mimetype='application/json')
+
+
+@blueprint.route('/<user_name>/taskbrowse_bookmarks/<short_name>/<bookmark_name>', methods=['DELETE'])
```

Comment:
How about making ```bookmark_name``` a body parameter then using the same endpoint as POST/GET? You could reuse a few lines of code. And more importantly, in case in the future, the DELETE API is expanded to do bulk delete or delete all.

Reply:
refactored in https://github.com/bloomberg/pybossa/pull/832/commits/7bafb0cc622d1cc1a44aed0572d1bd003967b9f5